### ISSUE TITLE
mqtttransportstage: _on_mqtt_disconnect called: Paho returned rc==0 caused by None

### ISSUE THREAD

**Issue by tosi2020** [Azure/azure-iot-sdk-python#454]
HI

I tried this python code on a raspberry pi 4, , when I send a message from vs code, the message is recieved at my raspberry, I get the output below but after some seconds also the error message:
Anyone knows how to solve this ?  (( had hoped that the program should wait for ever for messages)

OUTPUT:
--------------------------------------------------
Starting the IoT Hub Python sample...
IoTHubDeviceClient waiting for commands, press Ctrl-C to exit
Message received
    Data: <<b'Tjohoo'>>
    Properties: {}
    Total calls received: 1
MQTTTransportStage: _on_mqtt_disconnect called: Paho returned rc==1 caused by None
MQTTTransportStage: disconnection was unexpected
Exception caught in background thread.  Unable to handle.
['azure.iot.device.common.transport_exceptions.ConnectionDroppedError: None caused by Paho returned rc==1 caused by None\n']
-------------------------------------------------

Python code:
----------------------------------------------------------------
import threading
import time
from azure.iot.device import IoTHubDeviceClient

CONNECTION_STRING = "hide this one :-) "

RECEIVED_MESSAGES = 0

def message_listener(client):
    global RECEIVED_MESSAGES
    while True:
        message = client.receive_message()
        RECEIVED_MESSAGES += 1
        print("Message received")
        print( "    Data: <<{}>>".format(message.data) )
        print( "    Properties: {}".format(message.custom_properties))
        print( "    Total calls received: {}".format(RECEIVED_MESSAGES))

def iothub_client_sample_run():
    try:
        client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)

        message_listener_thread = threading.Thread(target=message_listener, args=(client,))
        message_listener_thread.daemon = True
        message_listener_thread.start()

        while True:
            time.sleep(1000)

    except KeyboardInterrupt:
        print ( "IoTHubDeviceClient sample stopped" )
        
if __name__ == '__main__':
    print ( "Starting the IoT Hub Python sample..." )
    print ( "IoTHubDeviceClient waiting for commands, press Ctrl-C to exit" )

    iothub_client_sample_run()
-----------------------------------------------

**Comment by david-davis** [comment]
I am also getting this same error, however
```
E paho: failed to receive on socket: [WinError 10054] An existing connection was forcibly closed by the remote host
E azure.iot.device.common.pipeline.pipeline_stages_mqtt: MQTTTransportStage: _on_mqtt_disconnect called: Paho returned rc==1 caused by None
W azure.iot.device.common.pipeline.pipeline_stages_mqtt: MQTTTransportStage: disconnection was unexpected
E azure.iot.device.common.handle_exceptions: Exception caught in background thread.  Unable to handle.
E azure.iot.device.common.handle_exceptions: ['azure.iot.device.common.transport_exceptions.ConnectionDroppedError: None caused by Paho returned rc==1 caused by None\n']
```
It happens only about once every 24 hours while running overnight tests on an 
IoTHubDeviceClient which synchronously calls send_message() on a separate
producer-consumer thread. The message rate is about 2 per second. 

The issue occurs on RPI Raspbian 10 *and* Windows 10 both running Py 3.7.3+, 
azure-iot-device 2.0.0, paho-mqtt 1.5.0.

Does anyone have a patch to handle the error with a reconnect?

**Comment by david-davis** [comment]
Found a fix in issue #399, with updated example code [call connect() method to prevent reconnect errors #18
](https://github.com/Azure-Samples/azure-iot-samples-python/pull/18)

**Comment by BertKleewein** [comment]
@tosi2020 - We're getting many false positives from this error, but the SDK is designed to recover from random disconnection.  We've changed the code (post 2.0.1) to record this at a lower logging level.  
It's unclear from your description if any functionality is broken or if you're just concerned because you saw the error output.   If the client stops receiving messages at this point, that may be a different issue.  @david-davis 's comment on #399 is also useful -- I have a fix that makes the workaround in #399 unnecessary in review #460.

Can you confirm if other functionality is broken when this happens, and if so, does adding a manual connect() call fixes your problem?

**Comment by LouanDuToitS3** [comment]
@BertKleewein Just to add some info. I have noticed that during the "disconnected" state that edgeHub logs "[WRN] - Module xxx/yyy is not connected" (where xxx/yyy is my edge device and the docker module name). Now in edgeHub v1.0.8.4 and before I used to get this continuesly, and the module will stay disconnected until I manually restart it. But sine v1.0.8.5 of edgeHub I see this being resolved by edgeHub logging "[INF] - Reauthenticating connected clients".

But once I still had a "disconnected" client which did not reconnect by itself, and was completely unaware if being disconnected. It also called "connect" before sending data. The only way to get it to be connected was to restart the module.

**Comment by BertKleewein** [comment]
@LouanDuToitS3 - I found a race condition in the code which could explain this.  If the SAS token gets renewed while the client is disconnected, after Paho sends a PINGREQ and before it times out for not receiving a PINGRESP, then the client will be disconnected but the Azure library will assume that it is still connected.  It's very unlikely, but I was able to repro it at least once.   I am working on a fix.

**Comment by LouanDuToitS3** [comment]
@BertKleewein Excellent, well done on finding it, looking forward to the fix, thanks.

**Comment by BertKleewein** [comment]
Fixed in 2.1.1.  If you see any new connection problems, please open a new issue.

**Comment by az-iot-builder-01** [comment]
@tosi2020, @david-davis, @BertKleewein, @LouanDuToitS3, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
